[DRAFT] Restore new semantics for ObjC metatype casts

tbkka opened the pull request:
This was originally merged in in PR#69435, but that had two serious problems that prompted it to be reverted in PR#71065:
* It unconditionally changed the semantics, which broke some existing code
* It forbade casting ObjC metatypes to NSObject, which is in fact valid

This is still WiP -- I've outlined the necessary backwards-compatibility logic to keep old code running, but still need to work through the ObjC metatype semantics.

tbkka committed:
[DRAFT] Restore new semantics for ObjC metatype casts

This was originally merged in in PR#69435, but that had two
serious problems that prompted it to be reverted in PR#71065:
* It unconditionally changed the semantics, which broke some existing code
* It forbade casting ObjC metatypes to NSObject, which is in fact valid

This is still WiP -- I've outlined the necessary backwards-compatibility
logic to keep old code running, but still need to work through the
ObjC metatype semantics.

tbkka commented:
@mikeash I'll need your help on the ObjC metatype semantics here.

tbkka committed:
Allow ObjC class objects to conform to NSObjectProtocol (but no other protocols)

mikeash commented:
We need to handle any protocol that the root class conforms to, not just the `NSObject` protocol.

I think the right way to handle this is to avoid using the `conformsToProtocol:` method, and instead use `class_conformsToProtocol`. That call doesn't check superclasses, so you need to do it in a loop, something like:

```
bool conforms = false;
for (Class c = object_getClass(object); c; c = class_getSuperclass(c)) {
  if (class_conformsToProtocol(c, proto)) {
    conforms = true;
    break;
  }
}
```

This is pretty much what the implementation of `-[NSObject conformsToProtocol:]` does. This will correctly handle class objects, because the loop will eventually check `NSObject` (or whatever the class's root class is).

This is a bit tricky because it's not clear what we want to do if someone overrides `conformsToProtocol:` to return something different. Should we believe it, or skip it, for Swift's purposes? I'm not sure. Changing our behavior will probably break something. We can continue calling the method for instances, and only do the loop for classes, or we could continue calling the method only in the bincompat scenario.

Just to describe what's going on a bit more:

1. ObjC classes are also ObjC objects.
2. The class of a class is its metaclass.
3. Metaclasses are also ObjC objects.
4. The class of a metaclass is the metaclass of the root class (i.e. the `NSObject` metaclass, for `NSObject` subclasses).
5. The class of the root class metaclass (argh) is the root class.
6. Result: a class is always an instance of its root class.
7. Result 2: the class, as an object, is an instance which conforms to all protocols that the root class conforms to.

In concrete terms:

```
@objc protocol P {}
extension NSObject : P {}

class Sub: NSObject {}

Sub.self is NSObject.Type // true, normal metatype subtype relationship
Sub.self is NSObject // also true, because classes are instances of their root class
Sub.self as AnyObject is P // true, because Sub.self is an NSObject instance, and NSObject conforms to P
```

There's a nice diagram here in case this description hasn't confused you enough already: http://sealiesoftware.com/blog/archive/2009/04/14/objc_explain_Classes_and_metaclasses.html

tbkka commented:
Ah.  Good point.  This was the part I had missed (that NSObject might conform to other protocols):
```
@objc protocol P {}
extension NSObject : P {}
```

tbkka commented:
@mikeash here's the test case I'm trying to get working.  Before I tear out any more hair, could you double-check that these are in fact the expected results (at least, following Obj-C semantics):
```
@objc protocol P1 {}
@objc protocol P2 {}
@objc class C: NSObject, P1 {}
extension NSObject: P2 {}

// Easy cases: instances follow inheritance rules
C() is any P1 // true
C() is NSObject // true
C() is NSObjectProtocol // true
C() is any P2 // true

// ObjC class objects inherit from NSObject
// just like everything else:
C.self is NSObject // true
C.self is any NSObjectProtocol // true
C.self is any P2 // true
C.self as AnyObject is any P2 // true

// But class object is not an instance of the class
C.self is any P1 // false

// Hmmm... Can we get ObjC metaclass objects in Swift?
```

tbkka commented:
Okay, I think I've gotten `swift_dynamicCastObjCProtocol*` matching the expectations above.  But now I see that when the optimizer calls `swift_dynamicCast` instead, it gets different results. So I'll have to sort that out as well.

tbkka committed:
Rework casting from metaclass to use a helper function

This seems to get us very nearly the expected results if we want
to see ObjC behaviors for ObjC metaclass objects when they
appear in Swift.

tbkka commented on stdlib/public/runtime/SwiftObject.mm:
```diff
@@ -1347,12 +1347,38 @@ static bool isObjCForUnownedReference(void *value) {
   return type;
 }
 
+static bool classObjectConformsToProtocol(id object, Protocol *protocol) {
```
This is basically copied from @mikeash's example in the discussion, but I had to do two `getClass` operations to make it work.  I'm not sure I understand that.

mikeash replied:
That doesn't look right. That will obtain the metaclass of `object` and then follow that superclass chain. `class_conformsToProtocol` should always say "no" here, until it gets to the root class and the weird stuff that happens there. Here's the implementation of `conformsToProtocol:` (taken from https://github.com/showxu/objc4/blob/75f57479d4416815ac43e81848db76d101f86eab/objc4/runtime/NSObject.mm#L2308):

```
- (BOOL)conformsToProtocol:(Protocol *)protocol {
    if (!protocol) return NO;
    for (Class tcls = [self class]; tcls; tcls = tcls->getSuperclass()) {
        if (class_conformsToProtocol(tcls, protocol)) return YES;
    }
    return NO;
}
```

`[self class]` will be equivalent to `object_getClass(self)` in the absence of any weird overrides.

mikeash replied:
If you're inspecting/debugging this stuff, it can be hard to tell apart classes and metaclasses, since they have the same name and the same `po` output in lldb. You can use `image lookup -va` and look for the symbol name (metaclasses will say `OBJC_METACLASS`) or call `class_isMetaClass(class)` to differentiate at runtime.

tbkka commented on test/Casting/Casts.swift:
```diff
@@ -1081,4 +1081,33 @@ CastsTests.test("type(of:) should look through __SwiftValue")
+  // ObjC metatypes:
+  expectTrue(C106973771.self is any NSObjectProtocol)
+  expectTrue(C106973771.self is NSObject.Type)
+  expectTrue(C106973771.self is NSObject) // FIXME
```
This line consistently fails.  I need to track down why; I suspect it might be that the compiler is optimizing this away differently than the runtime.

mikeash commented:
Your test cases look right to me. You can't directly get metaclasses in Swift using syntax like `C.self`, but you can get one with something like `object_getClass(C.self)`. Metaclasses will follow the same rules as classes (they are classes, after all), so they'll conform to protocols on their root class, and no others.

tbkka committed:
Implement ObjC metaclass awareness for general dynamic casting

Expanding the tests to force use of the most-general runtime casting
entrypoint uncovered some more places that need to understand this.
